File: interflow-frontend/src/model/besoin.ts
import { Matiere, MatiereModel } from './matiere';

export enum Etat {
  INCONNU = "inconnu",
  PARTIEL = "partiel",
  COUVERT = "couvert",
  NON_COUVERT = "non_couvert"
}

export interface Besoin {
  id?: string;
  matiere: Matiere;
  quantite: number;
  echeance: Date;
  etat: Etat;
  lot: string;
}

export class BesoinModel implements Besoin {
  id?: string;
  matiere: Matiere;
  quantite: number;
  echeance: Date;
  etat: Etat;
  lot: string;

  constructor(data: Partial<Besoin>) {
    this.matiere = data.matiere || new MatiereModel({ code_mp: "UNKNOWN", nom: "Matière inconnue" });
    this.quantite = data.quantite || 0;
    this.echeance = data.echeance || new Date();
    this.etat = data.etat || Etat.INCONNU;
    this.lot = data.lot || "";
    
    // Générer l'id à partir du code_mp de la matière + échéance + lot si non fourni
    if (!data.id && this.matiere) {
      const dateStr = this.echeance.toISOString().split('T')[0].replace(/-/g, '');
      const lotStr = this.lot.trim();
      if (lotStr) {
        this.id = `${this.matiere.code_mp}_${dateStr}_${lotStr}`;
      } else {
        this.id = `${this.matiere.code_mp}_${dateStr}`;
      }
    } else {
      this.id = data.id;
    }
  }

  static fromData(data: Record<string, unknown>): BesoinModel {
    const processedData = { ...data };
    
    // Gestion des valeurs par défaut pour les champs obligatoires
    if (!processedData.etat) {
      processedData.etat = Etat.INCONNU;
    }
    
    if (!processedData.lot) {
      processedData.lot = "";
    }
    
    // Si etat est une string, la convertir en enum
    if (typeof processedData.etat === 'string') {
      try {
        processedData.etat = Object.values(Etat).find(e => e === processedData.etat) || Etat.INCONNU;
      } catch (error) {
        processedData.etat = Etat.INCONNU;
      }
    }
    
    // Gestion de la matière si c'est un objet
    if (processedData.matiere && typeof processedData.matiere === 'object') {
      processedData.matiere = MatiereModel.fromData(processedData.matiere as Record<string, unknown>);
    }
    
    // Gestion de l'échéance si c'est une string
    if (processedData.echeance && typeof processedData.echeance === 'string') {
      try {
        processedData.echeance = new Date(processedData.echeance);
      } catch (error) {
        processedData.echeance = new Date();
      }
    }
    
    // On ne passe jamais id, il sera généré automatiquement
    delete processedData.id;
    
    try {
      return new BesoinModel(processedData);
    } catch (error) {
      console.error("Erreur lors de la création du besoin:", error);
      // En dernier recours, créer un besoin avec des valeurs par défaut
      return new BesoinModel({
        matiere: new MatiereModel({ code_mp: "UNKNOWN", nom: "Matière inconnue" }),
        quantite: 0,
        echeance: new Date(),
        etat: Etat.INCONNU,
        lot: ""
      });
    }
  }

  toData(): Record<string, unknown> {
    return {
      id: this.id,
      matiere: this.matiere instanceof MatiereModel ? this.matiere.toData() : this.matiere,
      quantite: this.quantite,
      echeance: this.echeance.toISOString(),
      etat: this.etat,
      lot: this.lot
    };
  }
} 

File: interflow-backend/src/models/besoin.py
from pydantic import BaseModel, Field
from models.matieres import Matiere

from datetime import datetime, timedelta
from enum import Enum
from typing import List, Iterator, Optional

class Etat(Enum):
    INCONNU = "inconnu"
    PARTIEL = "partiel"  # Couverture partielle
    COUVERT = "couvert"
    NON_COUVERT = "non_couvert"  # Ajout pour la cohérence


class Besoin(BaseModel):
    id: Optional[str] = Field(None, description="Identifiant unique du besoin (généré automatiquement)")
    matiere: Matiere = Field(..., description="Matière associée au besoin")
    quantite: float = Field(..., description="Quantité nécessaire", ge=0)
    echeance: datetime = Field(..., description="Date d'échéance du besoin")
    etat: Etat = Field(..., description="État du besoin")
    lot: str = Field(..., description="Numéro de lot")

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        # Générer l'id à partir du code_mp de la matière + échéance + lot si non fourni
        if not self.id and self.matiere:
            date_str = self.echeance.strftime("%Y%m%d")
            lot_str = self.lot.strip() if self.lot else ""
            if lot_str:
                self.id = f"{self.matiere.code_mp}_{date_str}_{lot_str}"
            else:
                self.id = f"{self.matiere.code_mp}_{date_str}"

    def model_dump(self):
        return {
            "id": self.id,
            "matiere": self.matiere.model_dump(),
            "quantite": self.quantite,
            "echeance": self.echeance.isoformat() if self.echeance else None,
            "etat": self.etat.value,
            "lot": self.lot
        }

    def model_dump_json(self, **kwargs):
        """
        Sérialise le modèle en JSON

        Args:
            **kwargs: Arguments passés à json.dumps

        Returns:
            str: Le JSON sérialisé
        """
        import json
        return json.dumps(self.model_dump(), **kwargs)

    @classmethod
    def from_model_dump(cls, data: dict):
        """
        Désérialise un dictionnaire en modèle Besoin

        Args:
            data: Le dictionnaire à désérialiser

        Returns:
            Besoin: L'instance de Besoin créée
        """
        # Copie du dictionnaire pour éviter de modifier l'original
        processed_data = data.copy()

        # Gestion des valeurs par défaut pour les champs obligatoires
        if "etat" not in processed_data or not processed_data["etat"]:
            processed_data["etat"] = Etat.INCONNU

        if "lot" not in processed_data or not processed_data["lot"]:
            processed_data["lot"] = ""

        # Si etat est une string, la convertir en enum
        if isinstance(processed_data["etat"], str):
            try:
                processed_data["etat"] = Etat(processed_data["etat"])
            except ValueError:
                processed_data["etat"] = Etat.INCONNU

        # Gestion de la matière si c'est un dictionnaire
        if "matiere" in processed_data and isinstance(processed_data["matiere"], dict):
            from models.matieres import Matiere
            processed_data["matiere"] = Matiere.from_model_dump(processed_data["matiere"])

        # Gestion de l'échéance si c'est une string
        if "echeance" in processed_data and isinstance(processed_data["echeance"], str):
            try:
                echeance = datetime.fromisoformat(processed_data["echeance"])
                # S'assurer que la date est naive (sans timezone) pour éviter les erreurs de comparaison
                if echeance.tzinfo is not None:
                    echeance = echeance.replace(tzinfo=None)
                processed_data["echeance"] = echeance
            except ValueError:
                # Fallback pour d'autres formats de date
                try:
                    processed_data["echeance"] = datetime.strptime(processed_data["echeance"], "%Y-%m-%d %H:%M:%S")
                except ValueError:
                    processed_data["echeance"] = datetime.now().replace(tzinfo=None)

        # On ne passe jamais id, il sera généré automatiquement
        processed_data.pop("id", None)

        # Création de l'instance
        try:
            return cls(**processed_data)
        except Exception as e:
            print(f"Erreur lors de la création du besoin: {e}")
            # En dernier recours, créer un besoin avec des valeurs par défaut
            from models.matieres import Matiere
            return cls(
                matiere=processed_data.get("matiere", Matiere(code_mp="UNKNOWN", nom="Matière inconnue")),
                quantite=processed_data.get("quantite", 0.0),
                echeance=processed_data.get("echeance", datetime.now().replace(tzinfo=None)),
                etat=processed_data.get("etat", Etat.INCONNU),
                lot=processed_data.get("lot", "")
            )
